src/backend/distributed/commands/index_pg_source.c

```diff
@@ -0,0 +1,195 @@
+/*-------------------------------------------------------------------------
+ *
+ * index_pg_source.c
+ *    Helper functions copy & pasted from PostgreSQL source code.
+ *    All the functions in this file is copied from
+ *    postgres/src/backend/commands/indexcmds.c
+ *
+ * Copyright (c) Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+
+#include "commands/defrem.h"
+#include "distributed/commands.h"
+#include "distributed/listutils.h"
+#include "mb/pg_wchar.h"
+
+
+/* *INDENT-OFF* */
+
+/*
+ * This function is copy & paste from Postgres source code:
+ * 		postgres/src/backend/commands/indexcmds.c
+ *
+ * Select the name to be used for an index.
+ *
+ * The argument list is pretty ad-hoc :-(
+ */
+char *
+ChooseIndexName(const char *tabname, Oid namespaceId,
+				List *colnames, List *exclusionOpNames,
+				bool primary, bool isconstraint)
+{
+	char *indexname;
+
+	if (primary)
```

Comment:
we cannot add tests, because we hit https://github.com/citusdata/citus/issues/1664. The reason we hit it is that this type of index only adds `table_name_pkey`. However, shard names are longer than `_pkey`.  We could add a test by manually setting shardId less than 4 chars, but that seems overkill